File: src/main/java/me/sargunvohra/leveluphp/capability/PlayerLevelStorage.java
package me.sargunvohra.leveluphp.capability;

import lombok.val;
import net.minecraft.nbt.INBTBase;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;

final class PlayerLevelStorage implements Capability.IStorage<PlayerLevelHandler> {
  private static final String XP_KEY = "xp";
  private static final String LEVEL_KEY = "level";

  @Override
  public INBTBase writeNBT(
      Capability<PlayerLevelHandler> type, PlayerLevelHandler instance, EnumFacing side) {
    val tag = new NBTTagCompound();
    tag.putInt(LEVEL_KEY, instance.getLevel());
    tag.putInt(XP_KEY, instance.getXp());
    return tag;
  }

  @Override
  public void readNBT(
      Capability<PlayerLevelHandler> type,
      PlayerLevelHandler instance,
      EnumFacing side,
      INBTBase nbt) {
    val tag = (NBTTagCompound) nbt;
    instance.setLevel(tag.getInt(LEVEL_KEY));
    instance.setXp(tag.getInt(XP_KEY));
  }
}

File: src/main/java/me/sargunvohra/leveluphp/core/PlayerLevelStorage.kt
package me.sargunvohra.leveluphp.core

import net.minecraft.nbt.INBTBase
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability

internal class PlayerLevelStorage : Capability.IStorage<PlayerLevelHandler> {

    override fun writeNBT(
        type: Capability<PlayerLevelHandler>,
        instance: PlayerLevelHandler,
        side: EnumFacing?
    ): INBTBase {
        val tag = NBTTagCompound()
        tag.putInt(LEVEL_KEY, instance.level)
        tag.putInt(XP_KEY, instance.xp)
        return tag
    }

    override fun readNBT(
        type: Capability<PlayerLevelHandler>,
        instance: PlayerLevelHandler,
        side: EnumFacing?,
        nbt: INBTBase
    ) {
        val tag = nbt as NBTTagCompound
        instance.level = tag.getInt(LEVEL_KEY)
        instance.xp = tag.getInt(XP_KEY)
    }

    companion object {
        private const val XP_KEY = "xp"
        private const val LEVEL_KEY = "level"
    }
}
